p2p/key.go

```diff
@@ -0,0 +1,122 @@
+package p2p
+
+import (
+	"bytes"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"math/big"
+
+	crypto "github.com/tendermint/go-crypto"
+	cmn "github.com/tendermint/tmlibs/common"
+)
+
+type ID string
+
+//------------------------------------------------------------------------------
+// Persistent peer ID
+// TODO: encrypt on disk
+
+// NodeKey is the persistent peer key.
+// It contains the nodes private key for authentication.
+type NodeKey struct {
+	PrivKey crypto.PrivKey `json:"priv_key"` // our priv key
+}
+
+// ID returns the peer's canonical ID - the hash of its public key.
+func (nodeKey *NodeKey) ID() ID {
+	return ID(hex.EncodeToString(nodeKey.id()))
+}
+
+func (nodeKey *NodeKey) id() []byte {
+	return nodeKey.PrivKey.PubKey().Address()
+}
+
+// PubKey returns the peer's PubKey
+func (nodeKey *NodeKey) PubKey() crypto.PubKey {
+	return nodeKey.PrivKey.PubKey()
+}
+
+func (nodeKey *NodeKey) SatisfiesTarget(target []byte) bool {
+	return bytes.Compare(nodeKey.id(), target) < 0
+}
+
+// LoadOrGenNodeKey attempts to load the NodeKey from the given filePath,
+// and checks that the corresponding ID is less than the target.
+// If the file does not exist, it generates and saves a new NodeKey
+// with ID less than target.
```

Comment:
Could you please clarify what `with ID less than target` means?

Reply:
From docs (https://github.com/tendermint/tendermint/blob/develop/p2p/docs/peer.md): Peer ID's must come with some Proof-of-Work; that is, they must satisfy peer.PrivKey.Address() < target for some difficulty target. This ensures they are not too easy to generate. To begin, let target == 2^240.

Reply:
Thanks @milosevic. However, that's still not disambiguated in prose. How about:
```go
// it generates and saves a new NodeKey whose ID's proof work is less than that of target
// thus ensuring that the ID is not too easy to generate.
``` 

What do you think of the above?

Reply:
@odeke-em It's more clear as you suggested; although we are dropping PoW for ID generation if I am not wrong.